ion-schema/src/constraint.rs

```diff
@@ -1031,6 +1067,117 @@ impl ConstraintValidator for FieldsConstraint {
     }
 }
 
+/// Implements an `field_names` constraint of Ion Schema
+/// [field_names]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#field_names
+#[derive(Debug, Clone, PartialEq)]
+pub struct FieldNamesConstraint {
+    type_reference: TypeReference,
+    requires_distinct: bool,
+}
+
+impl FieldNamesConstraint {
+    pub fn new(type_reference: TypeReference, requires_distinct: bool) -> Self {
+        Self {
+            type_reference,
+            requires_distinct,
+        }
+    }
+}
+
+impl ConstraintValidator for FieldNamesConstraint {
+    fn validate(
+        &self,
+        value: &IonSchemaElement,
+        type_store: &TypeStore,
+        ion_path: &mut IonPath,
+    ) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // create a set for checking duplicate elements
+        let mut element_set = vec![];
+
+        // get elements for given container in the form (ion_path_element, element_value)
+        let elements: Vec<(IonPathElement, &Element)> = match value {
+            IonSchemaElement::SingleElement(element) => {
+                // Check for null struct
+                if element.is_null() {
+                    return Err(Violation::new(
+                        "field_names",
+                        ViolationCode::TypeMismatched,
+                        "expected a struct but found null",
+                        ion_path,
+                    ));
+                }
+
```

Comment:
We can reduce the levels of nesting (a readability improvement, IMO) by using early returns in an `if let ... else` block. E.g.
```rust
let element = if let SingleElement(e) = value { e } else {
    return Err(Violation::new(
        "field_names",
        ViolationCode::TypeMismatched,
        format!("expected a struct but found document"),
        ion_path,
    ))
}

let the_struct = if let Some(s) = element.as_struct() { s } else {
    return Err(Violation::new(
        "field_names",
        ViolationCode::TypeMismatched,
        format!("expected a struct but found {}", element.ion_type()),
        ion_path,
    ))
}
```

Reply:
Using match arms with condition to improve readability.

Reply:
The match arms with condition eliminates one layer of the nesting/indentation, but it's not quite what I was suggesting.

I had forgotten that there is a `expect_element_of_type` function, but it is more along the lines of what I would like, and in fact, it's even better than what I originally suggested. You should do here what you did with the fields constraint validator. I.e.:
```rust
let ion_struct = value
    .expect_element_of_type(&[IonType::Struct], "field_names")?
    .as_struct()
    .unwrap();
```